core-bundle/src/Controller/ContentElement/DownloadsController.php

```diff
@@ -0,0 +1,243 @@
+                'file_previews' => $this->getPreviews($filesystemItem, $model),
+            ],
+            iterator_to_array($filesystemItems)
+        );
+
+        // Explicitly define title/text metadata for a single file
+        if ('download' === $model->type && $model->overwriteLink && $downloads) {
+            $downloads[0]['title'] = $model->linkTitle;
+            $downloads[0]['text'] = $model->titleText;
+        }
+
+        $template->set('downloads', $downloads);
+
+        return $template->getResponse();
+    }
+
+    /**
+     * Retrieve selected filesystem items but filter out those, that do not
+     * match the current DCA and configuration constraints.
+     */
+    private function getFilesystemItems(ContentModel $model): FilesystemItemIterator
+    {
+        $sources = match (true) {
+            'download' === $model->type => [$model->singleSRC],
+            $model->useHomeDir && ($user = $this->security->getUser()) instanceof FrontendUser && $user->assignDir && ($homeDir = $user->homeDir) => $homeDir,
+            default => $model->multiSRC,
+        };
+
+        // Find filesystem items
+        $filesystemItems = FilesystemUtil::listContentsFromSerialized($this->filesStorage, $sources);
+
+        // Optionally filter out files without metadata
+        if ('downloads' === $model->type && $model->metaIgnore) {
+            $filesystemItems = $filesystemItems->filter(
+                static fn (FilesystemItem $item): bool =>
+                    /** @var MetadataBag|null $metadata */
+                    null !== ($metadata = $item->getExtraMetadata()['metadata'] ?? null) &&
+                    null !== $metadata->getDefault()
+            );
+        }
+
+        // Only allow certain file extensions
+        $getAllowedFileExtensions = function (): array {
+            if (null !== $this->allowedFileExtensions) {
+                return $this->allowedFileExtensions;
+            }
+
+            $this->initializeContaoFramework();
+
+            return StringUtil::trimsplit(',', $this->getContaoAdapter(Config::class)->get('allowedDownload'));
+        };
+
+        return $filesystemItems->filter(
+            static fn (FilesystemItem $item): bool => \in_array(
+                Path::getExtension($item->getPath(), true),
+                array_map('strtolower', $getAllowedFileExtensions()),
+                true
+            )
+        );
+    }
+
+    /**
+     * If the content should be displayed inline or if the resource does not
+     * have a public URI, a URL pointing to this controller's download action
+     * will be generated, otherwise the direct download URL will be returned.
+     */
+    private function generateDownloadUrl(FilesystemItem $filesystemItem, ContentModel $model): string
+    {
+        $path = $filesystemItem->getPath();
+        $inline = $model->inline;
+
+        if (!$inline && null !== ($publicUri = $this->filesStorage->generatePublicUri($path))) {
```

Comment:
I think we should not use the public URI only if `!$inline` as the webserver might still deliver the response with content disposition inline...

Instead I would add a `contentDisposition` option to the `generatePublicUri` method (this is also supported by services like S3 (`ResponseContentDisposition`)).

Reply:
see 32015fe9e52aae1319df5fe0b39cf260ce9f9bf6